jvm/src/main/java/com/microsoft/bond/protocol/FastBinaryReader.java

```diff
@@ -95,70 +92,219 @@ public short readInt16() throws IOException {
-        return reader.readInt64();
+        return this.reader.readInt64();
     }
 
     @Override
     public byte readUInt8() throws IOException {
-        return reader.readInt8();
+        // reinterpret the sign bit as the high-order bit
+        return this.reader.readInt8();
     }
 
     @Override
     public short readUInt16() throws IOException {
-        return reader.readInt16();
+        // reinterpret the sign bit as the high-order bit
+        return this.reader.readInt16();
     }
 
     @Override
     public int readUInt32() throws IOException {
-        return reader.readInt32();
+        // reinterpret the sign bit as the high-order bit
+        return this.reader.readInt32();
     }
 
     @Override
     public long readUInt64() throws IOException {
-        return reader.readInt64();
+        // reinterpret the sign bit as the high-order bit
+        return this.reader.readInt64();
     }
 
     @Override
     public float readFloat() throws IOException {
-        return reader.readFloat();
+        return this.reader.readFloat();
     }
 
     @Override
     public double readDouble() throws IOException {
-        return reader.readDouble();
+        return this.reader.readDouble();
     }
 
     @Override
-    public byte[] readBytes(int count) throws IOException {
-        return reader.readBytes(count);
+    public byte[] readBytes(final int count) throws IOException {
+        return this.reader.readBytes(count);
     }
 
     @Override
     public boolean readBool() throws IOException {
-        return reader.readBool();
+        return this.reader.readBool();
     }
 
     @Override
     public String readString() throws IOException {
-        final int count = reader.readVarUInt32();
-        final byte[] bytes = reader.readBytes(count);
-        return StringHelper.decodeString(bytes);
+        final int codeUnitCount = reader.readVarUInt32();
+        if (codeUnitCount == 0) {
+            // avoid calling decoder for an empty byte array
+            return "";
+        } else {
+            final byte[] bytes = reader.readBytes(codeUnitCount);
+            return StringHelper.decodeString(bytes);
+        }
     }
 
     @Override
     public String readWString() throws IOException {
-        final int count = reader.readVarUInt32();
-        final byte[] bytes = reader.readBytes(count * 2);
-        return StringHelper.decodeString(bytes);
+        final int codeUnitCount = reader.readVarUInt32();
+        if (codeUnitCount == 0) {
+            // avoid calling decoder for an empty byte array
+            return "";
+        } else {
+            final byte[] bytes = reader.readBytes(codeUnitCount * 2);
+            return StringHelper.decodeWString(bytes);
+        }
     }
 
     @Override
-    public void skip(BondDataType type) throws IOException {
-        throw new RuntimeException("Not implemented.");
+    public void skip(final BondDataType type) throws IOException {
```

Comment:
getFixedTypeWidth is only for primitives while this function is for all types. I prefer to keep them separate for now.

---
In reply to: [119243379](https://github.com/Microsoft/bond/pull/465#discussion_r119243379) [](ancestors = 119243379)